flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java

```diff
@@ -1176,7 +1176,11 @@ private void allocateSlot(
     @Override
     public CompletableFuture<Acknowledge> freeSlot(
             AllocationID allocationId, Throwable cause, Time timeout) {
-        freeSlotInternal(allocationId, cause);
+        // only respond to freeing slots when not shutting down to avoid freeing slot allocation
+        // information
+        if (isRunning()) {
```

Comment:
Should we guard `#freeInactiveSlots` in a similar way?

I'm a bit miffed that we can get messages in the first place when the RpcEndpoint has stopped, but that would require bigger changes I guess.

Does the JM even need to try to release the slots if the TM disconnects? If the connection is closed doesn't the TM clean things up in any case right away?

Reply:
The behaviour is currently that the JM releases the TM if it gets a `disconnectTaskExecutor` message. This will then automatically free the slots. This can also be changed but is by far the more involved change because we have to distinguish whether the JM wants to release the slots or not in case of a `TaskExecutor` disconnect. I think there are also other cases when the JM wants to disconnect from a `TaskExecutor` (e.g. when itself shuts down). In this case you want to release the slots.

Probably it is a good idea to also guard the `freeInactiveSlots` call.

Yes, it is not trivial only allowing internal messages to be accepted during the shut down phase of an `RpcEndpoint`. That's why we introduced this `isRunning` method.